runtime/doc/news.txt

```diff
@@ -133,6 +133,8 @@ The following new APIs and features were added.
     `vim.treesitter.language.register`.
   • The `#set!` directive now supports `injection.self` and `injection.parent` for injecting either the current node's language
     or the parent LanguageTree's language, respectively.
+  • Added `vim.treesitter.preview_query()`, for live editing of treesitter
```

Comment:
"preview" was also shot down previously in the diagnostics API: https://github.com/neovim/neovim/pull/16057#issuecomment-945192362

Reply:
It's not too late to change, but I vehemently veto `edit` here. It's the wrong connotation **in this context**.

Reply:
.... `playground` ... 

...🍿

Reply:
> but I vehemently veto `edit` here. It's the wrong connotation **in this context**.

I understand the objection. But I don't understand why it's so strong. It's very close to how `:edit` works and is the entrypoint for many plugins with temporary buffers. 

I guess "scratch" at least draws from precedent. But still, if `:edit` is so bad, we need something long-term that describes the general case of a REPL/prompt use-case. "debug" is another idea.

Reply:
The objection comes from the context here, since we already have a bunch of functions acting on `query` -- which is always understood to be _the_ query loaded and applied to the buffer. 

This is a very different kind of query here, as it's ephemeral and not part of the "global state". 

`query_repl` would be suitable (even though it's technically not a REPL).

Reply:
FWIW I also prefer "edit" as the verb here. After all, the feature itself is the query _editor_. I'm trying to understand your objection @clason. 

>The objection comes from the context here, since we already have a bunch of functions acting on `query` -- which is always understood to be _the_ query loaded and applied to the buffer.

I am not sure what you mean by this. A buffer can have multiple queries.

>This is a very different kind of query here, as it's ephemeral and not part of the "global state". (You can have edit, but then you can't have query.) The same objection applies to debug_query.

Can you explain why you think the ephemeral nature of the query precludes `edit`? Whether a query is associated with a file on disk or lives in memory does not seem very relevant to me. It's still a query in every sense that matters (if it walks like a duck...).

The `vim.treesitter.query` module has methods both for "runtime queries" (those which exist as files on `'runtimepath'`) and for "ephemeral" queries (e.g. `vim.treesitter.query.parse` or `vim.treesitter.query.set`). So I don't think there is a strong precedent for "vim.treesitter.query implies a pre-existing, runtime query".

Reply:
Simple: `vim.treesitter.edit_query` implies (to me, obviously) that this will allow me to edit **the** (currently loaded, no matter where from) query.

Reply:
I have no preference here so just ping me when y'all have decided on the name and I'll change it 🙃

Reply:
> Simple: `vim.treesitter.edit_query` implies (to me, obviously) that this will allow me to edit **the** (currently loaded, no matter where from) query.
> 
> 
> Maybe to drive home the point better: why don't `query.edit()` and `query.get()` apply to the same query?

There is no singular query though. It is incorrect to refer to "**the** query" for a buffer, because a buffer can have multiple active queries. `vim.treesitter.query.get()` has a required argument for the query name (e.g. `vim.treesitter.query.get("c", "highlights")`. So a theoretical `vim.treesitter.query.edit` would have a similar interface. And the interface is extensible to "if `{query_name}` is omitted, edits a new, unnamed query".

For the purpose of this PR that is the only functionality that would be supported, because editing an existing query quickly runs into the difficulty of _which_ query should be edited (i.e. which file does `vim.treesitter.query.edit("c", "highlights")` actually edit?). But that can be solved another time.

> (Also, why is it not called `vim.treesitter.query.edit()` then -- with even stronger connotations in this regard?)

If you'll recall, [I proposed exactly this](https://github.com/neovim/neovim/pull/24703#discussion_r1300226016) 😇 .

Reply:
> If you'll recall, https://github.com/neovim/neovim/pull/24703#discussion_r1300226016 😇 .

Yes, and I liked this even less ;)

> editing an existing query quickly runs into the difficulty of which query should be edited (i.e. which file does vim.treesitter.query.edit("c", "highlights") actually edit?)

Yes, and that's why I want to steer as far away from that idea as possible.

Reply:
>Yes, and that's why I want to steer as far away from that idea as possible.

The ability to edit an existing query would be very useful, I would like to see that added Someday™️, but I agree that the "how" is unclear. But I do not agree that this is a use case we should never support, so taking baby steps _now_ to align us in that direction in the future makes sense to me.

Reply:
Well, that's now a different argument. If you envision that function to (eventually) do what I described as "false expectation", then the objection no longer applies.

As to how: The first step would be to simply `query.get()` the current effective query, put it in a scratch buffer, and load that as new query. You could add the information which files it was concatenated from -- if still available -- as a top comment, but I don't see how you could expect that simply `:w`riting it should work (without providing a path you are expected to have figured out).

The bigger question is how this should work with non-highlight queries...

Reply:
> `open_query_live_editor` would also be fine (but a mouthful).

"editor" is in that name :)

Does "inspect query" work? Either "edit query" or "inspect query" must work. There is no point in hyper-describing stuff in command names, that is not useful to users. Having generic verbs that _generally_ are understood to generally either "open stuff" or "develop stuff" is the main bifurcation. Over-specifying command names like `open_query_live_editor` kills discoverability. Users don't care about the pedantic details of whether "edit foo" precisely matches the case of "edit bar". They just want common entrypoints.

Our common entrypoints are "inspect" and "edit". If we identify a new, common paradigm that needs a name then we can introduce a new verb. But that isn't the case here.

I prefer "edit" but "inspect" is acceptable. Are there any objections to `inspect_query` ?

Reply:
If the goal is to extend this function to editing the "actual" queries (i.e., those returned by `query.get()`), then it should be `vim.treesitter.query.edit()`.

Reply:
Okay so it sounds like y'all agreed on "edit query"? If so, I'll change this from `InspectQuery` to `EditQuery`.

Also seems that I should be moving the implementation to the `query` module so that we have `vim.treesitter.query.edit` (?).

Reply:
No, I'd leave the implementation in the `dev` module and just expose it through `query.lua` instead of `treesitter.lua`.